File: src/lib/parse-agent.ts
import type { Block } from "./types";

const TOOL_RE =
  /<tool\s+name="([^"]+)"([^>]*)>([\s\S]*?)<\/tool>/gi;
const THINK_RE = /<think>([\s\S]*?)<\/think>/gi;

function attrs(src: string) {
  const path = /path="([^"]*)"/.exec(src)?.[1];
  const command = /command="([^"]*)"/.exec(src)?.[1];
  const to =
    /(?:\bto|\bsession|\bagent)="([^"]+)"/.exec(src)?.[1] ??
    /(?:\bto|\bsession|\bagent)=([^\s>"']+)/.exec(src)?.[1] ??
    path;
  return { path, command, to };
}

function pushText(blocks: Block[], text: string) {
  const t = text.replace(/\n{3,}/g, "\n\n").trim();
  if (t) blocks.push({ type: "text", text: t });
}

export function parseAgentOutput(raw: string): Block[] {
  const strippedThink: { text: string; thinks: string[] } = {
    text: raw,
    thinks: [],
  };
  strippedThink.text = raw.replace(THINK_RE, (_, body: string) => {
    strippedThink.thinks.push(String(body).trim());
    return "";
  });

  const blocks: Block[] = [];
  for (const t of strippedThink.thinks) {
    if (t) blocks.push({ type: "think", text: t });
  }

  const src = strippedThink.text;
  let last = 0;
  const re = new RegExp(TOOL_RE.source, "gi");
  let m: RegExpExecArray | null;
  while ((m = re.exec(src))) {
    pushText(blocks, src.slice(last, m.index));
    const a = attrs(m[2] ?? "");
    const name = m[1] ?? "Tool";
    blocks.push({
      type: "tool",
      name,
      path: a.path,
      command: a.command,
      to: /^sendmessage$/i.test(name) ? a.to : undefined,
      content: (m[3] ?? "").trim(),
      status: "done",
    });
    last = m.index + m[0].length;
  }

  const rest = src.slice(last);
  const open = /<tool\s+name="([^"]+)"([^>]*)>([\s\S]*)$/i.exec(rest);
  if (open && !/<\/tool>/i.test(rest)) {
    pushText(blocks, rest.slice(0, open.index));
    const a = attrs(open[2] ?? "");
    const name = open[1] ?? "Tool";
    blocks.push({
      type: "tool",
      name,
      path: a.path,
      command: a.command,
      to: /^sendmessage$/i.test(name) ? a.to : undefined,
      content: (open[3] ?? "").trim(),
      status: "running",
    });
  } else {
    pushText(blocks, rest);
  }

  return blocks.length ? blocks : [{ type: "text", text: raw.trim() }];
}

export function blocksToPlain(blocks: Block[]): string {
  return blocks
    .map((b) => {
      if (b.type === "text") return b.text;
      if (b.type === "think") return b.text;
      if (/^sendmessage$/i.test(b.name)) {
        return `SendMessage → ${b.to ?? ""}\n${b.content}`.trim();
      }
      const target = b.path ?? b.command ?? "";
      return `${b.name} ${target}\n${b.content}`.trim();
    })
    .join("\n\n");
}

/** Trailing text after the last tool/think is the reply; everything before is work. */
export function splitActivity(blocks: Block[]): {
  activity: Block[];
  reply: Block[];
} {
  let cut = -1;
  for (let i = 0; i < blocks.length; i++) {
    if (blocks[i]?.type !== "text") cut = i;
  }
  if (cut < 0) return { activity: [], reply: blocks };
  return {
    activity: blocks.slice(0, cut + 1),
    reply: blocks.slice(cut + 1),
  };
}

export function activitySummary(blocks: Block[]): string {
  const tools = blocks.filter((b) => b.type === "tool").length;
  const thinking = blocks.some((b) => b.type === "think");
  const parts: string[] = [];
  if (tools) parts.push(tools === 1 ? "1 tool" : `${tools} tools`);
  if (thinking) parts.push("thinking");
  return parts.join(" · ") || "activity";
}

export function lastSnippet(blocks: Block[], max = 140): string {
  const { activity, reply } = splitActivity(blocks);
  const prefer = reply.length ? reply : activity;
  for (const group of [prefer, blocks]) {
    for (let i = group.length - 1; i >= 0; i--) {
      const b = group[i];
      if (!b) continue;
      if (b.type === "text" && b.text.trim()) {
        const t = b.text.replace(/\s+/g, " ").trim();
        return t.length > max ? `${t.slice(0, max)}…` : t;
      }
      if (b.type === "tool") {
        if (/^sendmessage$/i.test(b.name)) {
          return `SendMessage · ${b.to ?? "session"}`;
        }
        const target = b.path ?? b.command ?? b.name;
        return `${b.name} · ${target}`;
      }
    }
  }
  return "No output yet";
}

export function extractSendMessages(blocks: Block[]) {
  // Any tool block whose name mentions westcode_send_message — the raw tool
  // id, an MCP-prefixed id, or a display title built from it — means the
  // message already went out over the MCP desk bus. Re-dispatching from the
  // XML tool blocks or the surrounding narration ("I sent it with
  // westcode_send_message to=grok …") would deliver the same task twice, so
  // this check runs before every extraction path.
  const deliveredViaMcp = blocks.some(
    (b) => b.type === "tool" && /westcode_send_message/i.test(b.name),
  );
  if (deliveredViaMcp) return [];

  const fromTools = blocks
    .filter(
      (b): b is Extract<Block, { type: "tool" }> =>
        b.type === "tool" && /^sendmessage$/i.test(b.name),
    )
    .map((b) => ({
      to: (b.to ?? b.path ?? "").trim(),
      text: b.content.trim(),
    }))
    .filter((s) => s.to && s.text);
  if (fromTools.length) return fromTools;

  const plain = blocksToPlain(blocks);
  const out: { to: string; text: string }[] = [];
  const xml =
    /<tool\s+name="SendMessage"\s+to="([^"]+)">([\s\S]*?)<\/tool>/gi;
  // No /m flag: with it, \s*$ matches at every line end and truncates a
  // multi-line body to its first line. Anchor to line starts by hand instead.
  const fence =
    /(?:^|\n)\s*westcode_send_message\s+(?:to[=:\s]+)([^\s\n]+)[\s\n]+([\s\S]+?)(?=\nwestcode_send_message|\s*$)/gi;
  const legacy =
    /(?:^|\n)\s*SendMessage\s+(?:to[=:\s"]+)([a-z0-9._-]+)["']?\s*\n+([\s\S]+?)(?=\nSendMessage\s+to|\s*$)/gi;
  let m: RegExpExecArray | null;
  for (const re of [xml, fence, legacy]) {
    re.lastIndex = 0;
    while ((m = re.exec(plain))) {
      out.push({ to: m[1]!.trim(), text: m[2]!.trim() });
    }
  }
  return out.filter((s) => s.to && s.text);
}

/** Render a session transcript as Markdown for export. */
export function exportTranscript(session: {
  title: string;
  providerId: string;
  model: string;
  cwd: string;
  createdAt: number;
  messages: {
    role: string;
    createdAt: number;
    blocks: Block[];
    fromTitle?: string;
    streaming?: boolean;
  }[];
}): string {
  const lines = [
    `# ${session.title}`,
    "",
    `Provider: ${session.providerId} · Model: ${session.model} · Folder: ${session.cwd}`,
    `Created: ${new Date(session.createdAt).toISOString()}`,
    "",
  ];
  for (const m of session.messages) {
    if (m.streaming || !m.blocks.length) continue;
    const who =
      m.role === "user"
        ? "User"
        : m.role === "assistant"
          ? "Assistant"
          : m.role === "agent"
            ? `Agent (${m.fromTitle ?? "peer"})`
            : "System";
    lines.push(`## ${who} — ${new Date(m.createdAt).toLocaleString()}`, "");
    if (m.role === "assistant") {
      const { activity, reply } = splitActivity(m.blocks);
      if (activity.length) {
        lines.push(
          "<details>",
          `<summary>${activitySummary(activity)}</summary>`,
          "",
        );
        writeBlocks(lines, activity);
        lines.push("</details>", "");
      }
      writeBlocks(lines, reply);
    } else {
      writeBlocks(lines, m.blocks);
    }
  }
  return lines.join("\n");
}

function writeBlocks(lines: string[], blocks: Block[]) {
  for (const b of blocks) {
    if (b.type === "text") lines.push(b.text, "");
    else if (b.type === "think") {
      lines.push(`> _${b.text.replaceAll("\n", " ")}_`, "");
    } else if (b.type === "tool") {
      lines.push(
        `**${b.name}**${b.path ? ` · ${b.path}` : ""}${b.command ? ` · \`${b.command}\`` : ""}`,
        "",
      );
      if (b.content) {
        // A fence longer than any run of backticks in the payload keeps
        // tool output that itself contains ``` from breaking the document.
        const runs = b.content.match(/`+/g) ?? [];
        const fence = "`".repeat(Math.max(3, ...runs.map((r) => r.length + 1)));
        lines.push(fence, b.content, fence, "");
      }
    }
  }
}

File: native/WestCode/WestCode/ParseAgent.swift
import Foundation

enum ParseAgent {
    static func parse(_ raw: String) -> [Block] {
        var thinks: [String] = []
        let thinkRe = try! NSRegularExpression(pattern: "<think>([\\s\\S]*?)</think>", options: .caseInsensitive)
        var ns = raw as NSString
        let full = NSRange(location: 0, length: ns.length)
        thinkRe.enumerateMatches(in: raw, options: [], range: full) { m, _, _ in
            guard let m, m.numberOfRanges > 1 else { return }
            thinks.append(ns.substring(with: m.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines))
        }
        var text = thinkRe.stringByReplacingMatches(in: raw, options: [], range: full, withTemplate: "")

        var blocks: [Block] = thinks.filter { !$0.isEmpty }.map { .think($0) }
        let toolRe = try! NSRegularExpression(
            pattern: "<tool\\s+name=\"([^\"]+)\"([^>]*)>([\\s\\S]*?)</tool>",
            options: .caseInsensitive
        )
        ns = text as NSString
        var last = 0
        toolRe.enumerateMatches(in: text, options: [], range: NSRange(location: 0, length: ns.length)) { m, _, _ in
            guard let m else { return }
            pushText(&blocks, ns.substring(with: NSRange(location: last, length: m.range.location - last)))
            let name = ns.substring(with: m.range(at: 1))
            let attr = ns.substring(with: m.range(at: 2))
            let body = ns.substring(with: m.range(at: 3)).trimmingCharacters(in: .whitespacesAndNewlines)
            let a = attrs(attr)
            blocks.append(.tool(.init(
                name: name, path: a.path, command: a.command,
                to: name.lowercased() == "sendmessage" ? a.to : nil,
                content: body, status: .done
            )))
            last = m.range.location + m.range.length
        }
        let rest = ns.substring(from: last)
        if let open = rest.range(of: #"<tool\s+name="([^"]+)"([^>]*)>([\s\S]*)$"#, options: .regularExpression),
           rest.range(of: "</tool>", options: .caseInsensitive) == nil {
            let before = String(rest[..<open.lowerBound])
            pushText(&blocks, before)
            let snippet = String(rest[open])
            let nameRe = try! NSRegularExpression(pattern: "name=\"([^\"]+)\"")
            let ns2 = snippet as NSString
            let name = nameRe.firstMatch(in: snippet, range: NSRange(location: 0, length: ns2.length))
                .map { ns2.substring(with: $0.range(at: 1)) } ?? "Tool"
            let a = attrs(snippet)
            let bodyStart = snippet.range(of: ">")?.upperBound
            let body = bodyStart.map { String(snippet[$0...]) }?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            blocks.append(.tool(.init(
                name: name, path: a.path, command: a.command,
                to: name.lowercased() == "sendmessage" ? a.to : nil,
                content: body, status: .running
            )))
        } else {
            pushText(&blocks, rest)
        }
        return blocks.isEmpty ? [.text(raw.trimmingCharacters(in: .whitespacesAndNewlines))] : blocks
    }

    static func blocksToPlain(_ blocks: [Block]) -> String {
        blocks.map { b -> String in
            switch b {
            case .text(let t), .think(let t): return t
            case .tool(let t) where t.name.lowercased() == "sendmessage":
                return "SendMessage → \(t.to ?? "")\n\(t.content)".trimmingCharacters(in: .whitespacesAndNewlines)
            case .tool(let t):
                return "\(t.name) \(t.path ?? t.command ?? "")\n\(t.content)".trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }.joined(separator: "\n\n")
    }

    static func lastSnippet(_ blocks: [Block], max: Int = 140) -> String {
        for b in blocks.reversed() {
            switch b {
            case .text(let t) where !t.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty:
                let s = t.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                return s.count > max ? String(s.prefix(max)) + "…" : s
            case .tool(let t) where t.name.lowercased() == "sendmessage":
                return "SendMessage · \(t.to ?? "session")"
            case .tool(let t):
                return "\(t.name) · \(t.path ?? t.command ?? t.name)"
            default: continue
            }
        }
        return "No output yet"
    }

    static func extractSendMessages(_ blocks: [Block]) -> [(to: String, text: String)] {
        let fromTools: [(to: String, text: String)] = blocks.compactMap { b in
            if case .tool(let t) = b, t.name.lowercased() == "sendmessage" {
                let to = (t.to ?? t.path ?? "").trimmingCharacters(in: .whitespaces)
                let text = t.content.trimmingCharacters(in: .whitespacesAndNewlines)
                if !to.isEmpty, !text.isEmpty { return (to, text) }
            }
            return nil
        }
        if !fromTools.isEmpty { return fromTools }
        let plain = blocksToPlain(blocks)
        let re = try! NSRegularExpression(
            pattern: #"SendMessage\s+(?:to[=:\s"]+)([a-z0-9._-]+)["']?\s*\n+([\s\S]+?)(?=\nSendMessage\s+to|\s*$)"#,
            options: .caseInsensitive
        )
        let ns = plain as NSString
        var out: [(to: String, text: String)] = []
        re.enumerateMatches(in: plain, range: NSRange(location: 0, length: ns.length)) { m, _, _ in
            guard let m, m.numberOfRanges > 2 else { return }
            let to = ns.substring(with: m.range(at: 1)).trimmingCharacters(in: .whitespaces)
            let text = ns.substring(with: m.range(at: 2)).trimmingCharacters(in: .whitespacesAndNewlines)
            if !to.isEmpty, !text.isEmpty { out.append((to, text)) }
        }
        return out
    }

    static func fillListAgents(_ blocks: [Block], roster: String) -> [Block] {
        blocks.map { b in
            if case .tool(let t) = b, t.name.lowercased() == "listagents" {
                return .tool(.init(name: t.name, path: t.path, command: t.command, to: t.to, content: roster, status: .done))
            }
            return b
        }
    }

    private static func pushText(_ blocks: inout [Block], _ text: String) {
        let t = text.replacingOccurrences(of: "\n{3,}", with: "\n\n", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if !t.isEmpty { blocks.append(.text(t)) }
    }

    private static func attrs(_ src: String) -> (path: String?, command: String?, to: String?) {
        func cap(_ pattern: String) -> String? {
            let re = try? NSRegularExpression(pattern: pattern)
            let ns = src as NSString
            guard let m = re?.firstMatch(in: src, range: NSRange(location: 0, length: ns.length)),
                  m.numberOfRanges > 1 else { return nil }
            return ns.substring(with: m.range(at: 1))
        }
        let path = cap("path=\"([^\"]*)\"")
        let command = cap("command=\"([^\"]*)\"")
        let to = cap("(?:\\bto|\\bsession|\\bagent)=\"([^\"]+)\"")
            ?? cap("(?:\\bto|\\bsession|\\bagent)=([^\\s>\"']+)")
            ?? path
        return (path, command, to)
    }
}
